File: app/rust/accounting-system/src/application/ports/output/audit_log_repository.rs
use async_trait::async_trait;
use crate::domain::audit::audit_log::AuditLog;

#[async_trait]
pub trait AuditLogRepository: Send + Sync {
    /// 監査ログを保存
    async fn save(&self, audit_log: AuditLog) -> Result<AuditLog, Box<dyn std::error::Error>>;

    /// エンティティの監査ログを取得
    async fn find_by_entity(
        &self,
        entity_type: &str,
        entity_id: &str,
        limit: i64,
        offset: i64,
    ) -> Result<Vec<AuditLog>, Box<dyn std::error::Error>>;

    /// ユーザーの監査ログを取得
    async fn find_by_user(
        &self,
        user_id: &str,
        limit: i64,
        offset: i64,
    ) -> Result<Vec<AuditLog>, Box<dyn std::error::Error>>;

    /// 全監査ログを取得（ページネーション）
    async fn find_all(
        &self,
        limit: i64,
        offset: i64,
    ) -> Result<Vec<AuditLog>, Box<dyn std::error::Error>>;
}

File: app/ruby/accounting_system/app/repositories/audit_log_repository.rb
# frozen_string_literal: true

class AuditLogRepository
  include AuditLogRepositoryInterface

  def save(audit_log)
    audit_log.save!
    audit_log
  end

  def find_by_entity(entity_type, entity_id)
    AuditLog.by_entity(entity_type, entity_id).to_a
  end

  def find_by_user(user_id, start_date, end_date)
    AuditLog.by_user(user_id).by_period(start_date, end_date).to_a
  end

  def find_by_period(start_date, end_date, limit)
    AuditLog.by_period(start_date, end_date).limit(limit).to_a
  end
end
